### DISCUSSION TITLE
Boolean型を外部ドメインクラスに指定したときの挙動について

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#933] [Q&A] [answered]
### 相談内容

いつもお世話になっております。
先日私の参加しているプロジェクトで、下記のような事象に遭遇しました。

1. Boolean型を外部ドメインとして指定し、DB上では数値で扱うように変換処理を入れる
2. EntityのフィールドにBoolean型のフィールドを持たせる
3. Criteria APIを使用し、whereで該当列に制約を設定する（`entityql.where(c -> c.eq(e.boolIntColumn, Boolean.TRUE))`）
4. 実行する
5. `org.postgresql.util.PSQLException: ERROR: 演算子が存在しません: smallint = boolean` が発生

このとき、プロジェクトのコンパイル自体は問題なく通ります。
左辺がsmallintで右辺がbooleanとなっているため、Entity_側の変換はうまく行っているものの、実際にバインドする値に対して外部ドメインの変換処理が上手く動いていないように見えました。

このような処理を内部ドメインクラスで実行すると、特に型変換等指定しなくてもスムーズに動いてくれているようです。
今のところBoolean以外の外部ドメインクラスについては試していないのですが、このような処理を行いたい場合、外部ドメインを指定することはできない、もしくは基本クラスとして扱われているようなクラスを外部ドメインに指定することは推奨されない、といった制約があるのでしょうか。
お忙しいところ大変恐縮ですが、ご教示頂けますと幸いです。

### 動作環境

+ Java Framework: Spring boot 2.7.4
+ DB: postres latest (15.2)
+ Doma2: 2.50.0

### 経緯

現在使用しているDBはPostgresですが、Oracleでの運用も考慮し、DBにboolean型の項目を持つのは極力避けたいと考えております。
そうした場合、Java上で真偽値として扱っている項目でもDB上は数値として扱う必要がありますが、DB上で数値１桁の項目として定義するだけでは、1もしくは0以外の値も想定されます。そのため、1もしくは0のどちらかの値のみがセットされる項目であることを、ドメインクラスを用いて表現するのが適切であると考えました。
数値の範囲を狭めることが目的であるため、「1」もしくは「0」のみを許容する内部ドメインクラスを作成しても良かったものの、真偽値であると認識できるものはJava上では直接Booleanで扱った方が良いのではないかという観点から、Booleanを外部ドメインとして指定する形で実装を進めていたところ、前述のエラーが発生しております。

このエラーが、プロジェクト側の設定・実装が問題なのか、なにかしら不具合によるものなのか、それとも仕様なのか判断がつかなかったため、ご相談させていただきました。

### 資産（簡略化しています）
<details><summary>table</summary>

```sql
create table xxx_table (
    key integer not null
  , value varchar(10)
  , bool_int smallint
  , primary key (key)
);

```
</details>

<details><summary>Entity</summary>

```java
@Entity
@Getter
public class XxxTable {
    @Id
    private Integer key;
    private String value;
    private Boolean boolInt;    // External domain class
}
```

</details>

<details><summary>Dao</summary>

```java
@Dao
public interface XxxTableDao {
    default List<XxxTable> method(Boolean bool) {
        XxxTable_ e = new XxxTable_();
        Entityql entityql = new Entityql(Config.get(this));
        return entityql
                .from(e)
                .where(c -> c.eq(e.boolInt, bool))    // org.postgresql.util.PSQLException
                .fetch();
    }
}
````

</details>

<details><summary>Converter</summary>

```java
@ExternalDomain
public class BoolIntConverter implements DomainConverter<Boolean, Integer> {
    @Override
    public Integer fromDomainToValue(Boolean value) {
        // Boolean を Integer に変換して返す
    }
    @Override
    public Boolean fromValueToDomain(Integer value) {
        // Integer を Boolean に変換して返す
    }
}
```

</details>
<details><summary>Provider</summary>

```java
@DomainConverters({
    BoolIntConverter.class
})
public class DomainConvertersProvider {
}
```

</details>

**Comment by nakamura-to** [comment, marked as answer]
詳細な共有ありがとうございます。

> もしくは基本クラスとして扱われているようなクラスを外部ドメインに指定することは推奨されない、といった制約があるのでしょうか。

はい、こちらに相当します。次のバージョンでは基本クラスを外部ドメインクラスとして扱おうとした場合にはエラーを出そうと思います。


「1」や「0」を真偽値として扱うには、基本データ型のマッピングをカスタマイズする方法を推奨します。以下ではその方法を示します。

まず、`PostgresDialect.PostgresJdbcMappingVisitor`を継承し、JDBCのマッピングにおいてBoolean型をInteger型で扱えるようにします。
```java
import org.seasar.doma.jdbc.JdbcMappingFunction;
import org.seasar.doma.jdbc.JdbcMappingHint;
import org.seasar.doma.jdbc.dialect.PostgresDialect;
import org.seasar.doma.jdbc.type.JdbcTypes;
import org.seasar.doma.wrapper.BooleanWrapper;

import java.sql.SQLException;

public class MyPostgresJdbcMappingVisitor extends PostgresDialect.PostgresJdbcMappingVisitor {

    @Override
    public Void visitBooleanWrapper(BooleanWrapper wrapper, JdbcMappingFunction p, JdbcMappingHint q) throws SQLException {
        return p.apply(wrapper, JdbcTypes.INTEGER_ADAPTIVE_BOOLEAN);
    }
}
```

次に、`PostgresDialect.PostgresSqlLogFormattingVisitor`を継承し、ログ上でBoolean型をInteger型に変換するようにします。（ログ上ではBoolean型のままがよければこのクラスの作成や設定は不要です）
```java
import org.seasar.doma.jdbc.SqlLogFormattingFunction;
import org.seasar.doma.jdbc.dialect.PostgresDialect;
import org.seasar.doma.jdbc.type.JdbcTypes;
import org.seasar.doma.wrapper.BooleanWrapper;

public class MyPostgresSqlLogFormattingVisitor extends PostgresDialect.PostgresSqlLogFormattingVisitor {

    @Override
    public String visitBooleanWrapper(BooleanWrapper wrapper, SqlLogFormattingFunction p, Void q) throws RuntimeException {
        return p.apply(wrapper, JdbcTypes.INTEGER_ADAPTIVE_BOOLEAN);
    }
}
```

[doma-spring-boot](https://github.com/domaframework/doma-spring-boot)を使っている想定で記載しますが、上記の2つのクラスを使って生成した`PostgresDialect`が使われるようにします。doma-spring-bootを使っていない場合には適宜読み替えてください。

```java
import org.seasar.doma.boot.autoconfigure.DomaConfigBuilder;
import org.seasar.doma.boot.autoconfigure.DomaProperties;
import org.seasar.doma.jdbc.dialect.Dialect;
import org.seasar.doma.jdbc.dialect.PostgresDialect;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class DomaConfiguration {

    @Bean
    public PostgresDialect dialect() {
        PostgresDialect dialect = new PostgresDialect(
            new MyPostgresJdbcMappingVisitor(),
            new MyPostgresSqlLogFormattingVisitor());
        return dialect;
    }

    @Bean
    public DomaConfigBuilder domaConfigBuilder(DomaProperties domaProperties, Dialect dialect) {
        DomaConfigBuilder builder = new DomaConfigBuilder(domaProperties);
        builder.dialect(dialect);
        return builder;
    }
}
```

上記の設定により、DB上の数値型をJava上の`Boolean`または`boolean`型にマッピングできます。

**Reply by ironclad-15256304** [reply]
迅速かつ丁寧なご回答有難うございます！
ご賢察の通りdoma-spring-bootを使用しております。
こちらの方でもご提示頂いた内容を参考に、対応してみたいと思います。

ちなみになんですが、基本クラスを外部ドメインクラスとして扱おうとした場合にエラーとする件については、何かこちらからIssueの発行等アクションを行ったほうが良いでしょうか？

**Reply by nakamura-to** [reply]
issueについては登録いただかなくても大丈夫です。ありがとうございます。

**Reply by backpaper0** [reply]
Dialectのbean定義さえあれば、domaConfigBuilderメソッドはなくても大丈夫です。
